nncf/torch/extensions/src/binarization/cuda/functions_cuda_impl.cu

```diff
@@ -3,27 +3,27 @@
 
 namespace {
 
-template <typename scalar_t>
+template <typename scalar_t, typename scalar_accum_t>
 __global__ void wb_cuda_scale_calc_kernel(
         const scalar_t* __restrict__ input,
         scalar_t* __restrict__ scale_output,
-        scalar_t* __restrict__ dev_tmp,
+        scalar_accum_t* __restrict__ dev_tmp,
         int* __restrict__ dev_last_block_counter,
         const int64_t total_elements_count) {
     const uint16_t tidx = threadIdx.x;
     const uint32_t bidx = blockIdx.x;
     const uint64_t gtidx = bidx * CUDA_MAX_NUM_THREADS_PER_BLOCK + tidx;
     const uint64_t grid_size = CUDA_MAX_NUM_THREADS_PER_BLOCK * gridDim.x;
 
-    scalar_t sum = 0;
+    scalar_accum_t sum = 0;
     for (int i = gtidx; i < total_elements_count; i += grid_size) {
         sum += abs(*(input + i));
     }
 
     sum /= total_elements_count;
 
-    __shared__ scalar_t sh_mem[CUDA_MAX_NUM_THREADS_PER_BLOCK];
-    reduce_with_shared_memory<scalar_t>(sh_mem, sum, tidx, bidx, dev_tmp, dev_last_block_counter, scale_output, gridDim.x);
+    __shared__ scalar_accum_t sh_mem[CUDA_MAX_NUM_THREADS_PER_BLOCK];
+    reduce_with_shared_memory<scalar_t, scalar_accum_t>(sh_mem, sum, tidx, bidx, dev_tmp, dev_last_block_counter, scale_output, gridDim.x);
```

Comment:
@ljaljushkin, I moved all FP16 accumulator to FP32, I did the same for the binarization kernels. Do I need to check this changes in some specific way?

Reply:
These questions is more to @vshampor 
I'd check sota eval for `resnet18_imagenet_binarization_xnor` and `resnet18_imagenet_binarization_dorefa` models.
Also, there's a `benchmark_binarize_layers.py`. Maybe it's worth checking the difference in performance